### PR TITLE
Adding infrastructure for input supervisor

### PR DESCRIPTION
**Pull request by yangwill**
PR adds the following features

- visualizer script that shows active channel and status of the input supervisor
- `ControllerFailureAggregator` class that aggregates error signals from other leaf systems in the controller diagram
- refactoring `InputSupervisor` to make the error checking more readable
- smaller changes that can be seen in the `integration_test.pmd` procman config

---
This change is [Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273)

### CODE REVIEW

**Review by yminchen — CHANGES_REQUESTED** [review]
> Reviewed 18 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273)* status: 18 of 19 files reviewed, 9 unresolved discussions (waiting on @yangwill and @yminchen)

___
*[examples/Cassie/BUILD.bazel, line 118 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqPmZpLETYiMg43V1JW:-MqPmZpLETYiMg43V1JX:b5vat2n) ([raw file](https://github.com/dairlab/dairlib/blob/23f742718f4b6231d9076e0543244f89a8b7318f/examples/Cassie/BUILD.bazel#L118)):*
> ```Starlark
>     ],
> )
> ```

Safer to leave the unit test 
___
*[examples/Cassie/dispatcher_robot_in.cc, line 150 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqPqthoBtW8Z5_SJXaZ:-MqPqthoBtW8Z5_SJXa_:blfr30j) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/dispatcher_robot_in.cc#L150)):*
> ```
> lcmt_robot_input
> ```
Why do we have a local lcm command publisher here? I think we have one in lcmdrivenloop already?
___
*[examples/Cassie/input_supervisor.cc, line 77 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqQMZbx5nMDorKHHVap:-MqQMZbx5nMDorKHHVaq:bw9l3os) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/input_supervisor.cc#L77)):*
> ```cpp
>   /// soft_estop: (operator_triggered) triggers when the soft-estop is engaged by the operator
>   /// is_nan: triggers whenever a NaN is received by the dispatcher to prevent sending bad motor commands
>   /// consecutive_failures: triggers whenever the velocity or actuator limits are reached
> ```

Seems to be a bug in naming 
___
*[examples/Cassie/input_supervisor.cc, line 159 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqQQEvQ2gqUn5bUvsNi:-MqQQEvQ2gqUn5bUvsNj:b42wus4) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/input_supervisor.cc#L159)):*
> ```
> command->get_value()
> ```
Looks like we need to use `output->get_data()` instead of `command->get_value`, because the former is filtered with effort limit
___
*[examples/Cassie/input_supervisor.cc, line 163 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqQPSRuDHex494-ikC8:-MqQPSRuDHex494-ikC9:b-gdukqr) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/input_supervisor.cc#L163)):*
> ```cpp
>             context.get_discrete_state(prev_efforts_index_).get_value();
>     output->SetDataVector(blended_effort);
>     if (fmod(command->get_timestamp(), 0.5) < 1e-4) {
> ```

What does this condition mean? Why 0.5?
___
*[examples/Cassie/input_supervisor.cc, line 244 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqQOWXK7wLFtziDksXa:-MqQOWXK7wLFtziDksXb:b-52db9) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/input_supervisor.cc#L244)):*
> ```cpp
>           discrete_state->get_mutable_value(prev_efforts_time_index_)[0] >
>       kMaxControllerDelay) {
>     discrete_state->get_mutable_value(prev_efforts_time_index_)[0] = 0.0;
> ```

I don't get the algorithm here. Why setting the previous time to 0?
___
*[multibody/multibody_utils.cc, line 140 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqPo4iiDZsvxSIfVntu:-MqPo4iiDZsvxSIfVntv:bfbps23) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/multibody/multibody_utils.cc#L140)):*
> ```cpp
>   // Add visual for the ground.
> //  plant->RegisterVisualGeometry(plant->world_body(), X_WG, HalfSpace(),
> //                                "visual");
> ```

Probably accidentally commented out
___
*[systems/controllers/controller_failure_aggregator.cc, line 47 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqPpUZXEsJXBAtuagp0:-MqPpUZXEsJXBAtuagp1:by4hz6f) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/systems/controllers/controller_failure_aggregator.cc#L47)):*
> ```
> lcmt_controller_fail
> ```
Is this saying that if we rank damping mode higher than complete shutdown? 
___
*[systems/controllers/osc/operational_space_control.cc, line 992 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqPpu0MAg7YGw-RSFwA:-MqPpu0MAg7YGw-RSFwB:b-6314rv) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/systems/controllers/osc/operational_space_control.cc#L992)):*
> ```cpp
>   output->set_timestamp(robot_output->get_timestamp());
>   if (context.get_time() > 5.0) {
>     output->get_mutable_value()(0) = 1.0;
> ```

Is this temporary code? It looks to me like we shut off the OSC after t = 5?

**Review by yangwill — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273)* status: 11 of 19 files reviewed, 9 unresolved discussions (waiting on @yangwill and @yminchen)

___
*[examples/Cassie/dispatcher_robot_in.cc, line 150 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqPqthoBtW8Z5_SJXaZ:-MqQSsqZ-H7As_slBewA:b-wh81hm) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/dispatcher_robot_in.cc#L150)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

> ```
> lcmt_robot_input
> ```
Why do we have a local lcm command publisher here? I think we have one in lcmdrivenloop already?
</blockquote></details>

`dispatcher_in` doesn't publish lcm command inputs, it publishes UDP messages to Cassie. However, if we want to test that on the Drake sim, we need to publish via lcm. This is why there is the `sim` flag
___
*[examples/Cassie/input_supervisor.cc, line 77 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqQMZbx5nMDorKHHVap:-MqQTwWZ3R73nP8Ktyf3:b-896fix) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/input_supervisor.cc#L77)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

Seems to be a bug in naming 
</blockquote></details>

Done.
___
*[examples/Cassie/input_supervisor.cc, line 159 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqQQEvQ2gqUn5bUvsNi:-MqQjdSz3Q3UhmfS0_oE:bd30ifx) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/input_supervisor.cc#L159)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

> ```
> command->get_value()
> ```
Looks like we need to use `output->get_data()` instead of `command->get_value`, because the former is filtered with effort limit
</blockquote></details>

good catch
___
*[examples/Cassie/input_supervisor.cc, line 163 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqQPSRuDHex494-ikC8:-MqQjVS68TzDHA9EnCAC:bkq1n1) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/input_supervisor.cc#L163)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

What does this condition mean? Why 0.5?
</blockquote></details>

oh this is just to print out every 0.5 seconds, instead of every iteration. We can probably remove it now that we know that blending efforts works
___
*[examples/Cassie/input_supervisor.cc, line 244 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqQOWXK7wLFtziDksXa:-MqQkPe44B8wjbQTZnBM:bx906d0) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/input_supervisor.cc#L244)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

I don't get the algorithm here. Why setting the previous time to 0?
</blockquote></details>

I wanted to make this error latching, but I should probably do that elsewhere.
___
*[multibody/multibody_utils.cc, line 140 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqPo4iiDZsvxSIfVntu:-MqQjz33EHISD8AlLeXo:b2r14xt) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/multibody/multibody_utils.cc#L140)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

Probably accidentally commented out
</blockquote></details>

I want to add a flag that disables visualizing the half plane. I'll make that a parameter

**Review by yangwill — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273)* status: 11 of 22 files reviewed, 9 unresolved discussions (waiting on @yminchen)

___
*[examples/Cassie/BUILD.bazel, line 118 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqPmZpLETYiMg43V1JW:-MqWH6Qu1OYR24XDj0EY:bamux83) ([raw file](https://github.com/dairlab/dairlib/blob/23f742718f4b6231d9076e0543244f89a8b7318f/examples/Cassie/BUILD.bazel#L118)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

Safer to leave the unit test 
</blockquote></details>

I put back the unit test
___
*[systems/controllers/controller_failure_aggregator.cc, line 47 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqPpUZXEsJXBAtuagp0:-MqWH9Dj3uWbeAuDox8h:brni8t6) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/systems/controllers/controller_failure_aggregator.cc#L47)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

> ```
> lcmt_controller_fail
> ```
Is this saying that if we rank damping mode higher than complete shutdown? 
</blockquote></details>

Yeah, soft-estop is the highest priority. I added comments that makes the hierarchy clearer
___
*[systems/controllers/osc/operational_space_control.cc, line 992 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqPpu0MAg7YGw-RSFwA:-MqWHEio5gJL3hD5LG3n:b6ugorj) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/systems/controllers/osc/operational_space_control.cc#L992)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

Is this temporary code? It looks to me like we shut off the OSC after t = 5? 
</blockquote></details>

Yes this should be removed before this PR is merged. This is just to simulate a failure

**Review by yangwill — COMMENTED** [review]
Added a backup signal to the `lcm_driven_loop`. The safety controller will trigger when the active controller has not sent a message for more than 0.1s.

> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273#-:-MqWHKGDC-i2Y9i2hWjf:bziee22)* status: 11 of 22 files reviewed, 9 unresolved discussions (waiting on @yminchen)

File: examples/Cassie/dispatcher_robot_in.cc
Comment (Brian-Acosta) [inline examples/Cassie/dispatcher_robot_in.cc:213]: We need to change this to not assume a hardcoded joint order before merging in the drake joint reorder

**Review by yangwill — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273)* status: 11 of 25 files reviewed, 10 unresolved discussions (waiting on @Brian-Acosta and @yminchen)

___
*[examples/Cassie/dispatcher_robot_in.cc, line 213 at r2](https://reviewable.io/reviews/dairlab/dairlib/273#-MqGoyekxsgyf0KGtBSb-r2-213:-Mtycgnq6X3k5RAqOl7x:b-2ksop5) ([raw file](https://github.com/dairlab/dairlib/blob/fe73555ad5ad8b1c405d457a4ffde6bb3e32d4f0/examples/Cassie/dispatcher_robot_in.cc#L213)):*
<details><summary><i>Previously, Brian-Acosta (Brian Acosta) wrote…</i></summary><blockquote>

We need to change this to not assume a hardcoded joint order before merging in the drake joint reorder 
</blockquote></details>

The PDConfigReceiver will map the joints to the correct actuators. Just tested with the newest version of drake.

**Comment by yangwill** [comment]
Just realized that the simulated encoders were not translated over correctly from MuJoCo, gonna update the PR

**Review by yminchen — CHANGES_REQUESTED** [review]
Reviewable only shows the full diff for me for some reason, so I didn't double check all the code. I looked through your replies to my previous comments and assumed you have addressed them correctly. 

> Reviewed 17 of 25 files at r8, 38 of 39 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273#-:-MxtgE2M86Hlqc8o1bTw:b9mvlk6)* status: 55 of 56 files reviewed, 2 unresolved discussions (waiting on @Brian-Acosta, @yangwill, and @yminchen)

___
*[examples/Cassie/input_supervisor.cc, line 244 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqQOWXK7wLFtziDksXa:-Mxtf5Q8AArsomOtIVzs:b-2rfkmf) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/input_supervisor.cc#L244)):*
<details><summary><i>Previously, yangwill (William Yang) wrote…</i></summary><blockquote>

I wanted to make this error latching, but I should probably do that elsewhere.
</blockquote></details>

Got it. Can you add a comment here? It will be easier to read

**Review by yminchen — APPROVED** [review]
LGTM after adding the comment in input_supervisor

> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273#-:-Mxth8vgAO72CeVS-f0O:b-b4mg0c)* status: 55 of 56 files reviewed, 2 unresolved discussions (waiting on @Brian-Acosta, @yangwill, and @yminchen)

**Review by Brian-Acosta — COMMENTED** [review]
> Reviewed 1 of 39 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273)* status: all files reviewed, 3 unresolved discussions (waiting on @Brian-Acosta and @yangwill)

___
*[multibody/multibody_utils.cc, line 140 at r9](https://reviewable.io/reviews/dairlab/dairlib/273#-MxtkUc58IvMEC3EAEW6:-MxtkUc58IvMEC3EAEW7:b-7rs5sm) ([raw file](https://github.com/dairlab/dairlib/blob/6c19c65b504d4ebc3c45a77c03fcf70a9d8cdc49/multibody/multibody_utils.cc#L140)):*
> ```cpp
> 
>   // Add visual for the ground.
> //  plant->RegisterVisualGeometry(plant->world_body(), X_WG, HalfSpace(),
> ```

Any reason this is commented out?

**Review by yminchen — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273)* status: all files reviewed, 3 unresolved discussions (waiting on @Brian-Acosta and @yangwill)

___
*[multibody/multibody_utils.cc, line 140 at r9](https://reviewable.io/reviews/dairlab/dairlib/273#-MxtkUc58IvMEC3EAEW6:-MxtsJ1K5yeAb6SV4_4o:bfwlow2) ([raw file](https://github.com/dairlab/dairlib/blob/6c19c65b504d4ebc3c45a77c03fcf70a9d8cdc49/multibody/multibody_utils.cc#L140)):*
<details><summary><i>Previously, Brian-Acosta (Brian Acosta) wrote…</i></summary><blockquote>

Any reason this is commented out?
</blockquote></details>

Oh Will said he will add a flag to make this optional, but looks like he missed it

**Review by yangwill — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/273)* status: 52 of 57 files reviewed, 3 unresolved discussions (waiting on @Brian-Acosta and @yminchen)

___
*[examples/Cassie/input_supervisor.cc, line 244 at r1](https://reviewable.io/reviews/dairlab/dairlib/273#-MqQOWXK7wLFtziDksXa:-MxuApYO0FT87x-J5jre:b-896fix) ([raw file](https://github.com/dairlab/dairlib/blob/198ef3d96586e7be48294e5d425be5967dc6f5a8/examples/Cassie/input_supervisor.cc#L244)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

Got it. Can you add a comment here? It will be easier to read 
</blockquote></details>

Done.
___
*[multibody/multibody_utils.cc, line 140 at r9](https://reviewable.io/reviews/dairlab/dairlib/273#-MxtkUc58IvMEC3EAEW6:-Mxu3rVO0xoPMJNm60S9:b-896fix) ([raw file](https://github.com/dairlab/dairlib/blob/6c19c65b504d4ebc3c45a77c03fcf70a9d8cdc49/multibody/multibody_utils.cc#L140)):*
<details><summary><i>Previously, yminchen (Yu-Ming Chen) wrote…</i></summary><blockquote>

Oh Will said he will add a flag to make this optional, but looks like he missed it
</blockquote></details>

Done.

**Pull request merged**
